crates/red_knot_python_semantic/resources/mdtest/narrow/truthiness.md

```diff
@@ -219,3 +219,49 @@ else:
 # TODO: It should be A. We should improve UnionBuilder or IntersectionBuilder. (issue #15023)
 reveal_type(y)  # revealed: A & ~AlwaysTruthy | A & ~AlwaysFalsy
 ```
+
+## Filtering in chained boolean expressions
+
+```py
+from typing import Literal
+
+class A: ...
+
+def _(x: Literal[0, 1]):
+    reveal_type(x or A())  # revealed: Literal[1] | A
+    reveal_type(x and A())  # revealed: Literal[0] | A
+
+# Filtering should not happen in cases where it makes the result type more complex
+def _(x: str):
+    reveal_type(x or A())  # revealed: str | A
+    reveal_type(x and A())  # revealed: str | A
+
+def _(x: bool | str):
+    reveal_type(x or A())  # revealed: Literal[True] | str | A
+    reveal_type(x and A())  # revealed: Literal[False] | str | A
+
+class Falsy:
+    def __bool__(self) -> Literal[False]: ...
+
+class Truthy:
+    def __bool__(self) -> Literal[True]: ...
+
+def _(x: Falsy | Truthy):
+    reveal_type(x or A())  # revealed: Truthy | A
+    reveal_type(x and A())  # revealed: Falsy | A
+
+class MetaFalsy(type):
+    def __bool__(self) -> Literal[False]: ...
+
+class MetaTruthy(type):
+    def __bool__(self) -> Literal[False]: ...
+
+class FalsyClass(metaclass=MetaFalsy): ...
+class TruthyClass(metaclass=MetaTruthy): ...
+
+def _(x: type[FalsyClass] | type[TruthyClass]):
+    # TODO: Should be `type[TruthyClass] | A`
+    reveal_type(x or A())  # revealed: type[FalsyClass] | type[TruthyClass] | A
+    # TODO: Should be `type[FalsyClass] | A`
+    reveal_type(x and A())  # revealed: type[FalsyClass] | type[TruthyClass] | A
```

Comment:
FWIW, in case you're interested, I don't think this should be a difficult TODO to address (and I think it will also fix some other TODOs in the tests.) It just requires handling `ClassLiteral` differently in `Type::bool`. (The TODO comment there says we should look up "`__bool__` and `__len__`" methods on the metaclass, but this is wrong, we've since realized we can't rely on `__len__` for anything, as explained below under the Instance case.)

In fact I think the simplest implementation might be to treat the `ClassLiteral` type as instead an `Instance` type of its metaclass, and just recursively defer to the `Instance` case.

Probably best to do this in a separate PR though.